vanetza/security/default_certificate_validator.cpp

```diff
@@ -0,0 +1,122 @@
+    // section 6.7 in TS 103 097 v1.2.1
+    unsigned certificate_time_constraints = 0;
+
+    // check validity restriction
+    for (auto& restriction : certificate.validity_restriction) {
+        ValidityRestriction validity_restriction = restriction;
+        ValidityRestrictionType type = get_type(validity_restriction);
+
+        if (type == ValidityRestrictionType::Time_Start_And_End) {
+            // change start and end time of certificate validity
+            StartAndEndValidity start_and_end = boost::get<StartAndEndValidity>(validity_restriction);
+
+            // check if certificate validity restriction timestamps are logically correct
+            if (start_and_end.start_validity >= start_and_end.end_validity) {
+                return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+            }
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            if (now < start_and_end.start_validity || now > start_and_end.end_validity) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_End) {
+            EndValidity end = boost::get<EndValidity>(validity_restriction);
+
+            // check if certificate is outdated
+            auto now = convert_time32(time_now);
+            if (now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_Start_And_Duration) {
+            StartAndDurationValidity start_and_duration = boost::get<StartAndDurationValidity>(validity_restriction);
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            std::chrono::seconds duration = start_and_duration.duration.to_seconds();
+            auto end = start_and_duration.start_validity + duration.count();
+            if (now < start_and_duration.start_validity || now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        }
+    }
+
+    // if not exactly one time constraint is given, we fail instead of considering it valid
+    if (1 != certificate_time_constraints) {
+        return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+    }
+
+    // check if subject_name is empty
+    if (0 != certificate.subject_info.subject_name.size()) {
+        return CertificateInvalidReason::INVALID_NAME;
+    }
+
+    // check signer info
+    if (get_type(certificate.signer_info) != SignerInfoType::Certificate_Digest_With_SHA256) {
+        return CertificateInvalidReason::INVALID_ROOT_HASH;
+    }
+
+    HashedId8 signer_hash = boost::get<HashedId8>(certificate.signer_info);
+    std::vector<Certificate> possible_signers = trust_store.find_by_id(signer_hash);
```

Comment:
Separate from that, I don't really understand how you're dealing with chains here. As far as I can tell this code assumes that the signer of the certificate currently being checked is already in the trust store (i.e., it's a root certificate). I don't think it is a good idea to cache intermediate CA certs in the trust store, unless you have some kind of periodic removal in place. But maybe I just misunderstood your code here.

Reply:
Yes, correct. Chain validation will follow in a separate PR and will likely require changes here. It's mostly required for the separation change that doesn't rely on a magic hash anymore.

Reply:
👍 awesome. It might be nice to have separate stores for the trusted certs and the "validation cache" (i.e., where previously validated intermediate certs, or even vehicle certs, can be cached).

Reply:
Yes, these will be separate, as root certs won't expire. But there might be an additional entity that combines both to make lookups easier.